File: java/MinimumPathSum.java
/*
 * Filename: MinimumPathSum.java
 * Author: Jun Xu
 * Email: jxu.vimacs@gmail.com
 * Created Time: Mon 18 May 2015 08:56:33 PM EDT
 */

public class Solution {
	public int minPathSum(int[][] grid) {
		int m = grid.length;
		int n = grid[0].length;
		int[] A = new int[n];

		for (int i = 0; i < n; i++) {
			if (i == 0) {
				A[i] = grid[0][i];
			} else {
				A[i] = A[i-1] + grid[0][i];
			}
		}

		for (int i = 1; i < m; i++) {
			for (int j = 0; j < n; j++) {
				if (j == 0) {
					A[j] += grid[i][j];
				} else {
					A[j] = Math.min(A[j-1], A[j]) + grid[i][j];
				}
			}
		}

		return A[n-1];
	}
}

File: cpp/minimum_path_sum.cc
/*
 * Filename: minimum_path_sum.cc
 * Author: Jun Xu
 * Email: jxu.vimacs@gmail.com
 * Created Time: Fri 31 Jul 2015 08:31:22 PM EDT
 */

#include <iostream>
using namespace std;

class Solution {
	public:
		int minPathSum(vector<vector<int> >& grid) {
			if (!grid.size() || !grid[0].size()) {
				return 0;
			}

			int m = grid.size();
			int n = grid[0].size();
			int dp[n];
			dp[0] = grid[0][0];

			for (size_t j = 1; j < n; j++) {
				dp[j] = dp[j-1] + grid[0][j];
			}

			for (size_t i = 1; i < m; i++) {
				dp[0] += grid[i][0];

				for (size_t j = 1; j < n; j++) {
					dp[j] = min(dp[j], dp[j-1]) + grid[i][j];
				}
			}

			return dp[n-1];
		}
};
